Global styles: block background UI controls

ramonjd opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Enables background images for supported blocks in global styles. 

https://github.com/WordPress/gutenberg/assets/6458278/c65a9deb-a1bb-41cd-8ebc-b9905f0735d9

Blocks include:

- https://github.com/WordPress/gutenberg/pull/62498
- https://github.com/WordPress/gutenberg/pull/62497
- https://github.com/WordPress/gutenberg/pull/62499

Part of:

- https://github.com/WordPress/gutenberg/issues/54336

Core backport PR:

- https://github.com/WordPress/wordpress-develop/pull/6836


## Why?

To enable background images across all supported blocks at once! MWHAHAHAHAHA!


## How?

Enabling background style properties as valid styles in `VALID_STYLES`, and not just valid `'top'` styles.

## Testing Instructions

To be updated...

In the Site Editor, open a template that contains a few ~Group~ Blocks.

Then head over to Styles > Blocks > ~Group~ and set a background image for all Group blocks. 

Check that all Group blocks now have that background image as default.

Ensure that the default background image is overridable at the block level. That is, select a Group block and set a different background image for that block.

Now define a background image at the block theme in theme.json.

Here's an example using some local files:

```json
{
	"$schema": "../../schemas/json/theme.json",
	"version": 3,
	"settings": {
		"appearanceTools": true
	},
	"styles": {
		"blocks": {
			"core/verse": {
				"background": {
					"backgroundImage": {
						"url": "file:./img/budgie4.jpg"
					}
				}
			},
			"core/quote": {
				"background": {
					"backgroundImage": {
						"url": "file:./img/budgie2.jpgg"
					}
				}
			},
			"core/pullquote": {
				"background": {
					"backgroundImage": {
						"url": "file:./img/budgie2.jpg"
					}
				}
			},
			"core/post-content": {
				"background": {
					"backgroundImage": {
						"url": "file:./img/budgie1.jpg"
					}
				}
			}
		}
	}
}

```

Check that this is displayed correctly in the frontend and editor.

ramonjd commented on schemas/json/theme.json:
```diff
@@ -2197,6 +2267,7 @@
 				},
 				{
 					"properties": {
+						"background": {},
```
TODO - double check all these properties. I YOLOed them.

ramonjd committed:
Adding background panel for blocks.

ramonjd committed:
Default controls using block supports

ramonjd commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -485,10 +485,10 @@ class WP_Theme_JSON_Gutenberg {
+			'backgroundImage'    => null,
+			'backgroundPosition' => null,
+			'backgroundRepeat'   => null,
+			'backgroundSize'     => null,
```
TODO

- add tests class-wp-theme-json-test.php
- double check that only blocks that have support for background image are included

ramonjd commented on schemas/json/theme.json:
```diff
@@ -1115,6 +1115,74 @@
+									"additionalProperties": false
+								},
+								{
+									"$ref": "#/definitions/refComplete"
```
Most of this is copy/paste, double check validity

ramonjd commented on packages/edit-site/src/components/global-styles/screen-block.js:
```diff
@@ -232,6 +241,19 @@ function ScreenBlock( { name, variation } ) {
 		setStyle( { ...newStyle, border: { ...updatedBorder, radius } } );
 	};
 
+	const onChangeBackground = ( newStyle ) => {
```
TODO: This is not required. We need to set the default-background size/position in the incoming style object if it's not already in `getStylesDeclarations` in packages/block-editor/src/components/global-styles/use-global-styles-output.js

```js
	blockStyles = {
		...blockStyles,
		background: {
			...blockStyles.background,
			...setBackgroundStyleDefaults( blockStyles.background ),
		}
	};
```

ramonjd replied:
Maybe in packages/block-editor/src/components/global-styles/use-global-styles-output.js

ramonjd replied:
See https://github.com/WordPress/gutenberg/pull/60405/files#diff-52acb55f718dce035b748387f60ef068cd3769674b553584f7df6f790e9fa84eR397

ramonjd committed:
Updating tests after rebase
Removed unrequired change function

ramonjd committed:
Set defaults for the global styles block group background image

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -391,6 +392,20 @@ export function getStylesDeclarations(
+			...blockStyles,
+			background: {
+				...blockStyles.background,
+				...setBackgroundStyleDefaults( blockStyles.background ),
```
Another approach happening in https://github.com/WordPress/gutenberg/pull/60578/files#diff-52acb55f718dce035b748387f60ef068cd3769674b553584f7df6f790e9fa84eR403

andrewserong replied:
Just double-checking, this comment is no longer relevant, is it? Looks like the approach here is that we're only applying the defaults to block-level in global styles 👍

ramonjd replied:
Yep, you're right. It was me cross linking and basically tying string to my fingers so I didn't forget 😄

ramonjd committed:
Enabling theme file resolution for blocks

ramonjd commented on lib/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -820,6 +821,33 @@ public static function get_resolved_theme_uris( $theme_json ) {
 		}
 
+		// Block styles.
+		if ( ! empty( $theme_json_data['styles']['blocks'] ) ) {
```
Probably should abstract this logic. There'll be elements and other things later I imagine.

Also add unit tests to cover this.

andrewserong replied:
What did you have in mind for the abstraction? A separate method, or function call to another file? Just curious how this part of the feature might evolve 🙂

tellthemachines replied:
Maybe once (if?) we support background images for elements, the whole `foreach` could be a function that takes an array of either blocks or elements so we can run it on both? I don't think it's worth abstracting until that happens though.

ramonjd replied:
>  I don't think it's worth abstracting until that happens though.

Good point. Let's cross that bridge when we come to it.

> What did you have in mind for the abstraction? A separate method, or function call to another file? Just curious how this part of the feature might evolve 

I haven't really thought about it, but the obvious target would be around building the resolved array items. So the if block starting from `wp_check_filetype()`. 

The future will involve target specific styles — top-level, blocks, elements and, eventually, variations, the latter two requiring possibly further nested loops. 

Maybe once elements comes along it'll be clearer how the abstraction should look.

ramonjd committed:
LINTO

ramonjd commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -4788,14 +4788,14 @@ public function test_get_shadow_styles_for_blocks() {
 	}
 
-	public function test_get_top_level_background_image_styles() {
+	public function test_get_top_level_and_block_background_image_styles() {
```
Note to self:

This is okay, but to specifically test block style output, I could test the output of `WP_Theme_JSON_Gutenberg::get_styles_for_block`, which is called (indirectly) by `WP_Theme_JSON_Gutenberg::get_stylesheet` anyway.

That means I keep `test_get_top_level_background_image_styles` and test top-level styles only, and create another test for blocks.
	
There's an effort to clean up tests, so let's wait for the outcome there:

- https://github.com/WordPress/wordpress-develop/pull/6734

ramonjd replied:
I'll wait until https://github.com/WordPress/gutenberg/pull/62463 is merged before rebasing and fixing these tests.

tellthemachines commented:
Just giving this a quick test run and noticed the global background image controls seem to have disappeared from the Layout section where they are on trunk. Did something go wrong in the rebase, or have the controls moved elsewhere?

Other than that, the controls seem to be working as expected! I'm loving the visual weirdness that comes from setting a global background image on the Group block (though I'm not sure how useful it's going to be in practice for Group - are we considering adding this as a feature to other blocks?)

<img width="1040" alt="Screenshot 2024-06-06 at 3 03 32 PM" src="https://github.com/WordPress/gutenberg/assets/8096000/a3e3c149-1ca3-4959-8237-f5afd85e774f">

ramonjd commented:
> Just giving this a quick test run and noticed the global background image controls seem to have disappeared from the Layout section where they are on trunk. Did something go wrong in the rebase, or have the controls moved elsewhere?

Thanks for testing @tellthemachines 

That's funky. I suspect human error, A.K.A I did something wrong 😄 I'll check it out in a bit.

andrewserong commented:
> (though I'm not sure how useful it's going to be in practice for Group - are we considering adding this as a feature to other blocks?)

I was wondering if it might be useful to add support to the Post Content block, thinking of sites that might want something like a parchment or paper texture for each post in a loop. But in that case, maybe the Post Template block could be a good candidate? Either way it might be tricky setting all instances of either block to have a background image, as the blocks might be used in different ways in different templates, but they could at least be interesting ones to play around with 🤔

ramonjd committed:
Use global settings to check if background image is supported.

ramonjd commented:
> Just giving this a quick test run and noticed the global background image controls seem to have disappeared from the Layout section where they are on trunk. Did something go wrong in the rebase, or have the controls moved elsewhere?

Ah I see... I copied the settings from `useSettingsForBlockElement` instead of the global settings. 

Since the background image in the layout panel doesn't really care about block supports I'll just pass it the raw global styles settings.

Fixed in [652d8ed](https://github.com/WordPress/gutenberg/pull/60100/commits/652d8edb1a27b1facebcef3290d77cf9a5612997)

andrewserong reviewed (commented):
This is testing pretty well for me so far! Defaults to `cover` at the block level in global styles, while site wide global styles continues to default to `tile` 👍

Continuing on from the conversation about the Group block and its usefulness when setting the background image for the block globally, I'm wondering if it'd be worth us trying out opting another block into the support so we've got a solid use case where setting it globally results in a pleasing looking site. Or to put it differently, can we try another block so that we can test whether or not this feature will feel usable for a real(ish) site design? I'm curious to prove out the idea of block-level in global styles, as we might find there are more blocks like Group where setting every block to a background image mightn't be all that desired. Could be something to ping designers or theme devs for feedback on?

Somewhat related, now that this PR enables setting all Group blocks to a particular background image, do we need the option to explicitly set to None at the individual block instance level in the editor? I.e. if I set a Group block to have a background image, how do I set an individual block not to have one:

<img width="1116" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/2cfb7bae-48ac-4170-a803-91a096fea931">

ramonjd commented:
> Continuing on from the conversation about the Group block and its usefulness when setting the background image for the block globally, I'm wondering if it'd be worth us trying out opting another block into the support so we've got a solid use case where setting it globally results in a pleasing looking site
> Could be something to ping designers or theme devs for feedback on?

I think containing blocks already unlock a lot of potential given the existing state of design controls.

Use cases for containing blocks could be layering (in the absence of _real_ layering), whereby group blocks are stacked or nested to create different effects, header and footer effects/

| Footer  | Stacked Group |
| ------------- | ------------- |
| <video src="https://github.com/WordPress/gutenberg/assets/6458278/4082efdb-dd33-4950-80f8-70332a81bc1a"/>  | <video src="https://github.com/WordPress/gutenberg/assets/6458278/034efc3f-368c-4a29-b7b5-8b5c87aa1c83"/>  |


Here's a nice example from https://www.parquesdesintra.pt/ that pertains to the Column block

<img width="600" alt="Screenshot 2024-06-11 at 4 14 03 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/68f5f268-dc84-4ff9-9d2a-0ec58d34fabe">


Keen to hear what @jasmussen and @jameskoster think too!

> Somewhat related, now that this PR enables setting all Group blocks to a particular background image, do we need the option to explicitly set to None at the individual block instance level in the editor?

Good point. I think allowing blocks to "opt-out" of any inherited background images would be useful.

I would expect any background images defined in theme.json for Group blocks, or any blocks, to display in the post editor. 

At the moment they don't, but thanks to https://github.com/WordPress/gutenberg/pull/57908, which exposes global styles in editor settings I think it's doable. I'll add an issue.

jasmussen commented:
Catching up. As best I can tell, the questions are: is it fine to add the background image control to more container blocks? And should it be allowed to set a default background to such a container block in global styles, requiring an explicit "none" for those same container blocks locally to override?

If I got that right, the answer is in the category of "yes and". Fundamentally, we want this for all design tools:

* Always available in global styles, always visible there, never hidden behind ellipsis/toolspanel functionality
* Always local opt-outs, in what form that might take
* Always local consideration for prominence, i.e. to show or not to show.

The nuance  here is that at the moment, Background Image is not just a design tool, it's a design tool _and a panel_. That means it's not something we can locally hide/reduce prominence of, should we want to. I don't know that we do want to hide that locally, maybe it's fine. But it's resurfacing a conversation we had in the past of: what's the best taxonomy for the background image design tool? Does it have to have a bespoke panel, or can it be grouped with other items? There's added nuance given the added tools that surface once you do add a background image—position, focal point, repeat, etc. Separately, there's the divergence with the Cover block, which has its own implementation of this. 

So we have a few tasks:

* Design a version of the Background image design tool that supports an explicit "no background".
* Design this in a way that the tool can absorb the legacy implementation currently present in the Cover block, unifying the interface across. 
* Consider the context of this tool: is a dedicated panel still the right location, or can it be absorbed in another? E.g. would it be crazy to rename the "Color" panel and include it there?

Strawman sketch:

<img width="484" alt="color and fill" src="https://github.com/WordPress/gutenberg/assets/1204802/8cee9081-db41-4695-bb55-6138daad500a">

ramonjd commented:
Thanks for the feedback @jasmussen 🙇🏻 🙇🏻 🙇🏻 

> Always available in global styles, always visible there, never hidden behind ellipsis/toolspanel functionality
> Always local opt-outs, in what form that might take
> Always local consideration for prominence, i.e. to show or not to show.

The local opt-out functionality I can do independently - the tricky part, as you allude to, will be first deciding on the taxonomy and they integrating it with current tools. 

> "Color & Fill", like we have "Border & Shadow" for the Button block.

Yes!

This is very timely, and I think will instruct the approach to another problem that background images has exposed:

- https://github.com/WordPress/gutenberg/issues/32787
- https://github.com/WordPress/gutenberg/issues/60401


From a CSS perspective, `color.gradient` and `background.backgroundImage` must be consolidated, preferably before 6.7. 

I think your idea addresses the UI part of the consolidation that, I think, must also occur given how they interplay. For example, a user adding a background gradient should see that it also affects the background image, that they can be overlayed, even [multiple values](https://github.com/WordPress/gutenberg/issues/60401#issuecomment-2034970543) applied.

tellthemachines commented:
> Use cases for containing blocks could be layering (in the absence of _real_ layering using the value of `background-image` and multiple gradient/images), whereby group blocks are stacked or nested to create different effects, header and footer effects/

I might be missing something here, but these examples all use either the sitewide background image or background images on specific blocks, whereas what I initially questioned was the use cases for setting a background image globally on all Group blocks (particularly because that includes all Row, Stack and Grid blocks too). I can't think of any case where you'd want to do that. It would be good to ship this functionality with at least one block where it'll actually come in useful 😅 

@andrewserong's suggestion of Post Content might be a good place to start! Text blocks used for particular effects like Quote and Verse might be good ones too.

jasmussen commented:
> I think your idea addresses the UI part of the consolidation that, I think, must also occur given how they interplay. For example, a user adding a background gradient should see that it also affects the background image, that they can be overlayed, even https://github.com/WordPress/gutenberg/issues/60401#issuecomment-2034970543 applied.

Without more coffee this morning I can't yet fully know that the ItemGroup as a stacking mechanism for layered backgrounds is right. It might be, or we might want a _single_ "Background" item where the stacking/management happens in the flyout. 

In either case, it may be worth looking at the overall effort of moving the contents of the current Background Image panel into such a flyout—same type of flyout as color and gradient presets currently live in. Hopefully some time next week I can get some time to visualize this a bit better, but sharing in case it resonates.

ramonjd commented:
> I can't yet fully know that the ItemGroup as a stacking mechanism for layered backgrounds is right.

Get your coffee! Ha. I don't think we need to expend thinking time on this right now, just to be aware that CSS allows it (and it's pretty cool). 

I'm pretty motivated by the "Color & Fill" idea alone as it unblocks an intellectual clog I've been having over how to package the merge of background gradients and background image 🌮

ramonjd committed:
This commit checks for theme block styles in the block supports hook so a background block control can determine whether it has an inherited value. If there is an inherited value, the block control will provide an option to "remove" the theme style.

This commit also sends resolved theme asset URI via the _links property so that these images can be used in block control previews.

Added resolver tests for blocks

Use `get_styles_for_block` in unit tests when testing block nodes

Use `get_styles_for_block` in unit tests when testing block nodes

Add since annotation

backport changelog

ramonjd commented on packages/editor/src/components/provider/use-block-editor-settings.js:
```diff
@@ -268,6 +273,7 @@ function useBlockEditorSettings( settings, postType, postId, renderingMode ) {
 				)
 			),
 			[ globalStylesDataKey ]: globalStylesData,
+			[ globalStylesLinksDataKey ]: globalStylesLinksData,
```
I need to catch up with discussion on https://github.com/WordPress/gutenberg/pull/61556, but I'm wary of adding another global styles property. 

I suppose it's safer than blindly passing on everything from `mergedGlobalStyles`, but wondering if they should be combined into a `{ styles, settings, _links }` object and we deprecate `__experimentalFeatures`?

andrewserong replied:
Just double-checking, is this a question that needs an answer before this PR can land? If so, is it worth pinging for wider feedback?

My understanding is that by using `globalStylesLinksDataKey` it's effectively private(ish) for now 🤔

ramonjd replied:
Good question, sorry I missed this last night.

It's private and so I'm okay with it 😄 I was mainly wondering if there were any concerns about overloading. 

I guess we'll ask for forgiveness 😇

youknowriad replied:
What is this "links" data here? It's very unclear to me. I don't think we should be adding random block editor settings without a good reflexion on these. 

Aside: the useMemo is missing a dependency.

ramonjd replied:
The value comprises resolved, relative paths to files defined in theme.json. 

For example, if `/my-theme/image.png` is referenced as a background image in theme.json, the global styles REST API will return the full URI to the file in the `_links` object so that the editor can correctly render the image. See https://github.com/WordPress/gutenberg/pull/61271

In that way, it forms part of "globalStylesData", and required for the reasons outlined in https://github.com/WordPress/gutenberg/pull/61556, which introduced "globalStylesData". 

So, unclear maybe, but not random in my opinion.

The `Links` in `globalStylesLinksData` is an extension of `globalStylesData`, which, is a fairly opaque label in its own right.

If the value is unclear, then I think we should rename it. What about:

- `globalStylesResolvedURIs`
- `themeAssetPaths`
- ?

Or merging all global styles data into one property?

> Aside: the useMemo is missing a dependency.

Whoops, thanks for spotting that! 🙇🏻 

Update here:

- https://github.com/WordPress/gutenberg/pull/63823

ramonjd replied:
> I suppose it's safer than blindly passing on everything from mergedGlobalStyles, but wondering if they should be combined into a { styles, settings, _links } object and we deprecate __experimentalFeatures?

There was also this question above, as I was concerned about adding another property.

youknowriad replied:
> The Links in globalStylesLinksData is an extension of globalStylesData, which, is a fairly opaque label in its own right.

I agree that `globalStylesData` is an opaque label but it has a clear purpose: We're introducing support for a "theme" to block-editor package basically. 

> For example, if /my-theme/image.png is referenced as a background image in theme.json, the global styles REST API will return the full URI to the file in the _links object so that the editor can correctly render the image. See https://github.com/WordPress/gutenberg/pull/61271

Why aren't we instead replacing the urls in the endpoint in `globalStylesData` directly instead of having a separate `_links` property? What do we do for fonts and things like that?

ramonjd replied:
Thanks for questioning this. I spent a lot of time to-ing and fro-ing, so I admit I could be blinkered on the subject.

> Why aren't we instead replacing the urls in the endpoint in globalStylesData directly instead of having a separate _links property? 

That's the first approach I took, but ran into some snags.

The main reason I [rejected](https://github.com/WordPress/gutenberg/pull/61271#issuecomment-2099776380) that approach was because:

- when switching to a theme style variations, its contents are saved in the user global CPT in the database
- should a user want to export the theme, the relative paths should be preserved to enable theme portability

Resolving the image URIs on the backend once means there's no switching between the two: between relative and absolute, which would be especially tricky if the theme is a child theme referring to assets in some parent theme.

> What do we do for fonts and things like that?

I did look into fonts briefly, and we're using the [same syntax](https://github.com/WordPress/gutenberg/blob/b3c120a7e157dfaadf36d6911bbf5a706351bfd3/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face-resolver.php#L148-L148) to represent.

What other assets work this way aside from fonts?

Also, I think fonts work differently in the way they're loaded. They're printed to the page once so their URLs can be resolved when printing out the [style tags](https://github.com/WordPress/gutenberg/blob/b3c120a7e157dfaadf36d6911bbf5a706351bfd3/lib/compat/wordpress-6.4/fonts/font-face/class-wp-font-face.php#L233-L233). 

Have the resolved background image URIs available in the frontend means we can apply them immediately without affecting the integrity of theme.json.

See for example how applying a variation that has a theme font needs a page reload to work:

https://github.com/user-attachments/assets/d9cce0d5-a9ed-4b71-be4b-3aedcb8cc7cd

cc @creativecoder for fact check

I'm happy to create an issue and investigate, if there's a specific issue to address here.

youknowriad replied:
> The main reason I https://github.com/WordPress/gutenberg/pull/61271#issuecomment-2099776380 that approach was because:
> when switching to a theme style variations, its contents are saved in the user global CPT in the database
should a user want to export the theme, the relative paths should be preserved to enable theme portability
Resolving the image URIs on the backend once means there's no switching between the two: between relative and absolute, which would be especially tricky if the theme is a child theme referring to assets in some parent theme.

While I'm not entirely convinced about the downsides, I think at least we should avoid a new key in the BlockEditorProvider setting and maybe resolve the images properly before setting `globalStylesData` even if it's done in the frontend.

Personally I still don't like that we need two different objects to represent `theme.json` or Global Styles. So far you just receive a simple object and it's enough to resolve everything.

ramonjd replied:
> resolve the images properly before setting globalStylesData even if it's done in the frontend.

So long as `globalStylesData` isn't merged into the current user style object, and then saved to the database, it should be fine to find the images and replace them with the resolved path in `_links`. 

I'll check, but have a feeling I tried this and rolled it because it _was_ merged to the user style object. 

Also, I'd have to go back and double check the performance hit since the styles are rendered whenever a global style is updated.

youknowriad replied:
> So long as globalStylesData isn't merged into the current user style object, and then saved to the database, it should be fine to find the images and replace them with the resolved path in _links.

Why is this such a big deal? I'm trying to understand, if you change a relative path with an absolute path, you're just making things clearer no?

ramonjd replied:
> Why is this such a big deal? I'm trying to understand, if you change a relative path with an absolute path, you're just making things clearer no?

It's only important if we care about preserving theme portability.

If an absolute path is saved to the database, then that absolute path will feature in the theme.json when a user exports their theme from the editor.

With the background image no longer referring to the exported theme's internal assets, it makes the exported theme very fragile and less portable. It places the onus on the user to dig into the JSON and fix the path.

I'm not dying on this hill by the way 😄 and am trying to think of alternatives, but I think they should satisfy the above requirement.

I also don't want to lose sight of the perceived area of concern — that editor settings has a new, private property that's related to the theme styles — and balancing the stakes here.
___ 

Sorry for the wall of text, but here's what I understand just to clarify:

1. Theme.json may refer to theme assets (images) using relative paths.
2.  The relative theme path could refer to a local asset, or a parent theme asset.
3. In order to build CSS in the editor (clientside) and on the frontend (PHP), we need to resolve the relative theme paths into absolute URLs.
4. The preference is to resolve the URLs on the backend using `get_theme_file_uri` as it will not only look at parent theme assets, but apply any filters.
5. Resolved URIs needs to be passed to the client (via REST) in order to build the correct CSS in the editor. 
6. To preserve theme.json style values, resolved URIs are embedded in `_links` in the response object.
7. In the client, the relative path -> absolute path substitution takes place when the CSS is built.
8. If a user applies a theme style variation that contains a relative path, that relative path is saved to the database, and follows the same process starting from point 1.

Now, with the way we've structured `_links` to include a theme.json path to the value, my guess is that it might be possible to replace the "urls in the endpoint in globalStylesData directly" as you mention, and then swap them back out before we commit any user styles to the global styles CPT in the database. To do this, we'd still need access to the `_links` object.

youknowriad replied:
I think you're right that the `_links` property make sense in the global styles endpoint. It gives the possibility to edit the styles while retaining the local aspect and add other local assets... This is something that can work conceptually.

I'm just thinking from the `block-editor` package's perspective, it might not make sense and I wonder if this discussion ties to the other discussion where I suggested as separate provider/package for Global Styles and whether that could simplify things here.

Obviously for now, this is private API, so it's "okey" but I would love if we can also follow-up with the right abstractions / right packages because we want third-parties to benefit from this as well.

---
Aside, I'm curious what's the shape of the `_links` object now?

ramonjd replied:
Great discussion, thanks for helping me think through it again. 

The `_links` object follows the documented standard for embedded URI resources:

https://developer.wordpress.org/rest-api/using-the-rest-api/linking-and-embedding/

```json
"_links": {
    "wp:theme-file-uris": [
        {
            "name": "file:/img/budgie3.jpg",
            "href": "http://ilovebudiges.com/wp-content/themes/emptytheme/img/budgie3.jpg",
            "target": "styles.background.backgroundImage.url",
            "type": "image/jpeg"
        }
    ]
}
```

"href", "type" and "name" are [HAL](https://en.wikipedia.org/wiki/Hypertext_Application_Language) reserved attributes. 

"target", a valid link attribute, represents the path in the theme.json tree to the value that needs to be replaced.

> follow-up with the right abstractions / right packages because we want third-parties to benefit from this as wel

The discussion over at https://github.com/WordPress/gutenberg/issues/63796 is sounding very productive. If I get down time during the next release cycle, I'd love to help out in any way.

ramonjd commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -134,10 +138,25 @@ export function BackgroundImagePanel( {
+				 * @TODO 1. Pass inherited value down to all block style controls,
+				 *   See: packages/block-editor/src/hooks/style.js
+				 * @TODO 2. Add support for block style variations,
+				 *   See implementation: packages/block-editor/src/hooks/block-style-variation.js
```
So if a variation has been applied to a block, we need to parse it here. 

This needs to happen in a follow up. I've added a note to https://github.com/WordPress/gutenberg/issues/54336

cc @aaronrobertshaw just in case there are gotchas (TL;DR - I'm trying to pass down the theme styles applied to a block to the controls, so that the controls are aware of any defaults)

aaronrobertshaw replied:
Yeah, at a glance, we'd need to detect which, if any, variation has been applied to the block, then merge the variation's data over the block type's data. The resulting set would tell us which value the control should get.

ramonjd commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -134,10 +138,25 @@ export function BackgroundImagePanel( {
+				style: getBlockAttributes( clientId )?.style,
+				_links: _settings[ globalStylesLinksDataKey ],
+				/*
+				 * @TODO 1. Pass inherited value down to all block style controls,
```
Probably doesn't need to happen any time soon, but might be useful for other block controls I imagine?

ramonjd commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -170,13 +189,15 @@ export function BackgroundImagePanel( {
 			settings={ updatedSettings }
 			onChange={ onChange }
 			value={ style }
+			themeFileURIs={ _links?.[ 'wp:theme-file' ] }
```
This is the magic that allows the background image preview to display where the image comes from theme.json

ramonjd reviewed (commented):
d782979 checks for theme block styles in the block supports hook so a background block control can determine whether it has an inherited value. If there is an inherited value, the block control will provide an option to "remove" the theme style.


It also sends resolved theme asset URI via the _links property so that these images can be used in block control previews.

https://github.com/WordPress/gutenberg/assets/6458278/0f3b9498-d59a-4a82-9e3f-468ab49e071c


> also in this PR, turn off global support for Group via the default theme.json (or however)

So I was a bit spaced out when I proposed this. I don't think it's a good idea for the plugin to turn off global styles support without allowing theme authors to override it via theme.json, which isn't straight forward at the moment.

I was chatting with @andrewserong and the standing opinion is that it's fine to leave support for Group block and just not promote it as an example, rather concentrating on the other, fine suggestions such as Verse et. al.

ramonjd committed:
update assertion comments

jameskoster commented:
It's not something to do here, but for a long time now I've been wondering whether color controls should appear in contextually relevant panels. IE move the Text color control to the Typography panel; move the Background color control to a new 'Background' or 'Fill' panel (alongside the background image control).

<img width="838" alt="bg" src="https://github.com/WordPress/gutenberg/assets/846565/56aa5e9f-462e-41e4-914d-60854f89b56e">


This is already the case for border color, and it works well imo:

<img width="280" alt="Screenshot 2024-06-19 at 15 10 40" src="https://github.com/WordPress/gutenberg/assets/846565/dc80c471-b38b-4a30-9ffb-345fa296de5b">

Users could re-arrange the order of backgrounds by dragging/dropping the itemgroup in the Background panel.

This could effectively replace the bespoke 'overlay' controls in the Cover block. To add an overlay you'd just add a background color, drag it above the background image/media and adjust the opacity. 

I don't think we should do stack arrangement in the flyout because eventually that's where the settings associated with the background image should live:

<img width="556" alt="Screenshot 2024-06-19 at 15 16 46" src="https://github.com/WordPress/gutenberg/assets/846565/a5b616af-0090-4382-b786-7671dca65691">

ramonjd commented:
Thanks @jameskoster !

I've picked up the following PR again, and am working on the fly out. 

- https://github.com/WordPress/gutenberg/pull/60151

I think I'll need some UX/UI help over there. 

I think, in terms of development, it'd be good to iterate first:

- on the existing controls in https://github.com/WordPress/gutenberg/pull/60151, e.g., fine tune and clean up the fly out taking into account all the controls, and any new ones (maybe we'll need tabs?) such as [further repeat options](https://github.com/WordPress/gutenberg/issues/62608) and [background attachment](https://github.com/WordPress/gutenberg/pull/61382)
- fix the [discord between gradient and background image](https://github.com/WordPress/gutenberg/pull/60739) and, parallel to that, 
- look at where background image controls should live in the UI

jasmussen commented:
> IE move the Text color control to the Typography panel; move the Background color control to a new 'Background' or 'Fill' panel (alongside the background image control).

That's an interesting thought. Would you be up for mocking up a few screens that go into edgecases here? Both across blocks that have _only_ the color panel at present (some templating blocks), or blocks where the color panel is used for something that isn't the text color, as well as the various background options? Another option I've thought about is this:

<img src="https://i0.wp.com/make.wordpress.org/design/files/2024/06/color-and-fill.png?w=968&ssl=1">

jameskoster commented:
Hmm yes there's actually more to the Color panel than I realised. For containers you can apply color styles to many child elements and blocks:

<img width="280" alt="Screenshot 2024-06-20 at 13 02 50" src="https://github.com/WordPress/gutenberg/assets/846565/69d6d40b-9566-4c42-9815-ffeff9caa3d6">

I find it a bit limiting that you can change the colors of these child elements, but nothing else. In any case, if those controls moved to the Typography and Background panels respectively we'd probably need to separate the elements somehow. As an extreme example, with _everything_ added the result could be something like this:

<img width="838" alt="colors" src="https://github.com/WordPress/gutenberg/assets/846565/af18adb9-5efe-4071-b2e5-ae639611d2a0">

Obviously this isn't ideal. However, to address the aforementioned limitations I think an 'Elements' panel could make more sense. This would allow you to drill down and style _all_ the properties of those elements for the selected container. 

<img width="937" alt="elements" src="https://github.com/WordPress/gutenberg/assets/846565/008852c9-28e6-4572-b57f-f012c3cd28ce">

This would be similar to how styling blocks works in Global styles.

jasmussen commented:
Solid points, Jay, solid mockups. There's some missing pieces to the mockups I've dabbled with [here](https://github.com/WordPress/gutenberg/pull/60151#issuecomment-2182198931), so please feel free to commandeer the Figma if some of your ideas connect. A few things do seem clear:

* Having the empty-state "Add background image" open the media library directly.
 * Also: avoid the dropdown unless a plugin is adding to it. It isn't helpful to have a dropdown that only shows "Open media library" and "Upload", since the former supports both already.
* Adding the minus reset button. Perhaps it only appears on hover. But this seems useful in _most_ itemgroups, even Color, Background, the list of fonts, etc.
* Having the position, tiling options, repeat, and other background image tools live in the flyout, where they can breathe, have helptext, etc, without encumbering the inspector any more. 

Now it's a matter of connecting those dots, and the primary question to me becomes: where does this Background Image button live? We can keep it in its own Background Image _panel_ as a start, if that helps PRs get moving. But it really fills duplicative to have it live there.

ramonjd commented:
Now that 

- https://github.com/WordPress/gutenberg/pull/60151

is merged, I've rebased this PR. I think it's up next to be 👀 and 🚢 

🙇🏻

ramonjd committed:
Now that background styles can be used in block styles, move them to styles complete ref

andrewserong reviewed (commented):
This is testing very well for me! I only ran into a couple of tiny issues so far:

When in the post editor and using the global styles value, the circular icon within the popover doesn't appear to render the preview image:

![image](https://github.com/WordPress/gutenberg/assets/14988353/993a2959-c4f6-4482-8d3e-648f0ac4d475)

When using the global styles value, at the individual block level, you're able to open up the popover and give different values for position, size, etc. However on the site frontend, these values are not output (I imagine because the block support thinks no image is set?). Should we allow overriding these individual values even if no image is set? Here's a screenshot to show what I mean:

| Block editor, updating size of global styles image in this particular instance | Site frontend, where the global styles image is output but the block level size is ignored |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/44a371d6-9990-4eb0-afc4-f0f3b70f9dca) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/82d956e7-43e1-4df4-8025-4508e7401f5c) |

Otherwise looking good:

✅ Theme relative urls are working nicely
✅ "Remove" option at the global styles level works
✅ Reset gets the `theme.json` value back again
✅ User uploaded values are working
✅ Site wide and block-level in post content background images are still working well as on `trunk`

So, overall this is feeling very close! For testing I mostly used the Quote and Pullquote blocks, and it's feeling nice to use to me 👍

ramonjd committed:
Pass resolved URIs to display inherited background image styles with relative paths
Also, output any background styles, even if there's no background image

ramonjd commented:
Thanks for the detailed review and feedback @andrewserong 

Fortunately those two bugs you caught were human (my) oversight. 

> When in the post editor and using the global styles value, the circular icon within the popover doesn't appear to render the preview image:

I wasn't passing the resolved theme paths 🤦🏻 

Fixed!

> Should we allow overriding these individual values even if no image is set? 

YES! I have fixed this. There was a check in background.php to only output these values if there was a background image, but I think now that we have support for block styles in theme.json (in this PR) we'll need to allow all properties through.

I've fixed that.

andrewserong reviewed (approved):
Thanks for the follow-up, that's all testing well for me now!

| Block level in post editor when global style is set | Block level sets size, image is from global style (frontend) |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/c453f2ad-2d56-4242-90ad-a8fbd575f69e) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/9c37804d-6aac-49bf-86ad-60ac9bafbfe3) |

The code change looks good to me, too. I see there was a question (https://github.com/WordPress/gutenberg/pull/60100#discussion_r1639127652) about the use of the `globalStylesLinksDataKey`. Since that's effectively private, I reckon this change is good to try for now, and we could always look at moving things around in follow-ups if need be?

LGTM! 🚀

jasmussen commented:
I've a few things today so just looking at your helpful screenshots (thanks Andrew); can the position thumbnail be the full width of the flyout?

ramonjd commented:
> I've a few things today so just looking at your helpful screenshots (thanks Andrew); can the position thumbnail be the full width of the flyout?

It _would_ look a lot prettier. 

There was a big of to and fro with this one over at https://github.com/WordPress/gutenberg/pull/60151#issuecomment-2204767101

I had to restrict the size of the image to ensure the popover height didn't blow out. 

The images can be very very short, or very very tall - it's not possible to know in advance, and allowing them to fill the width ends up with things like this:

<img width="719" alt="Screenshot 2024-07-05 at 5 44 59 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/b51c56b3-c338-423c-811a-3fe541ff6643">

The only alternative I can think of right now is to do some JS math to get the scaled ratio of w/h and then enforce a limit if the ratio is over the threshold of what looks weird, technically speaking 😄

jasmussen commented:
At worst, we can make the parent container full-width, and throw a gray background color inside, so it's an image inside a frame.  That would also help with the super tall thin ones, then the image can be scaled down to be small inside such a gray container.

ramonjd commented:
Thanks for the help and feedback everyone!

> That would fit in a full-width gray box with a max-height of, say, 180px. I'm sure there's all sorts of complexity in getting that to work, though!

Great idea @jasmussen !  I'll have a play around in a follow up. 🙇🏻 

Follow ups:

- look at [variation styles](https://github.com/WordPress/gutenberg/pull/60100#discussion_r1666141389)
- keep an eye on how the image preview in the focal picker behaves
